test/hotspot/jtreg/compiler/c2/irTests/XorINodeIdealizationTests.java

```diff
@@ -332,4 +342,105 @@ public boolean testFoldableXor(int x, int y) {
         var xor = (x & 0b111) ^ (y & 0b100);
         return xor < 0b1000;
     }
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testFoldableXorDifferingLength(int x, int y) {
+        var xor = (x & 0b111) ^ (y & 0b11);
+        return xor < 0b1000;
+    }
+
+    @Test
+    public int testXorMax(int x, int y) {
+        return (x & MAX_VALUE) ^ (y & 0b11);
+        // can't do the folding range check here since xor <= MAX_VALUE is
+        // constant with or without the xor
+    }
+
+    private static final Range RANGE_1 = Range.generate(G.restricted(0, MAX_VALUE));
+    private static final Range RANGE_2 = Range.generate(G.restricted(0, MAX_VALUE));
+    private static final int UPPER_BOUND = Integer.max(0, Integer.highestOneBit(RANGE_1.hi() | RANGE_2.hi()) * 2 - 1);
+
+    private static final int LIMIT_1 = G.next();
+    private static final int LIMIT_2 = G.next();
+    private static final int LIMIT_3 = G.next();
+    private static final int LIMIT_4 = G.next();
+    private static final int LIMIT_5 = G.next();
+    private static final int LIMIT_6 = G.next();
+    private static final int LIMIT_7 = G.next();
+    private static final int LIMIT_8 = G.next();
+
+
+    @Test
+    @IR(failOn = {IRNode.XOR})
+    @IR(counts = {IRNode.CON_I, "1"})
+    public boolean testFoldableRange(int x, int y) {
+        return (RANGE_1.clamp(x) ^ RANGE_2.clamp(y)) <= UPPER_BOUND;
+    }
+
+    @Test
+    public int testRandomLimits(int x, int y) {
+        x = RANGE_1.clamp(x);
+        y = RANGE_2.clamp(y);
```

Comment:
Question:
did you verify that this `RANGE_1` with its clamp values are really detected as constants by C2, and not seen as loads? Maybe that works, but I've never tried it myself.

Reply:
I  just hope that the abstraction here does not invalidate our intent to have constant min/max bounds for the clamping ;)

Reply:
Ah, you are right, great :)